src/v/iceberg/table_update_applier.cc

```diff
@@ -0,0 +1,172 @@
+// Copyright 2024 Redpanda Data, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+#include "iceberg/table_update_applier.h"
+
+#include "base/vlog.h"
+#include "container/chunked_hash_map.h"
+#include "container/fragmented_vector.h"
+#include "iceberg/logger.h"
+#include "iceberg/schema.h"
+#include "iceberg/table_metadata.h"
+#include "iceberg/table_update.h"
+
+namespace iceberg::table_update {
+
+namespace {
+
+struct update_applying_visitor {
+    explicit update_applying_visitor(table_metadata& meta)
+      : meta(meta) {}
+    table_metadata& meta;
+
+    outcome operator()(const add_schema& update) {
+        auto sid = update.schema.schema_id;
+        auto s = std::find_if(
+          meta.schemas.begin(),
+          meta.schemas.end(),
+          [sid](const schema& schema) { return schema.schema_id == sid; });
+        if (s != meta.schemas.end()) {
+            vlog(log.error, "Schema {} already exists", sid);
+            return outcome::unexpected_state;
+        }
+        if (update.last_column_id.has_value()) {
+            auto new_last_col_id = update.last_column_id.value();
+            if (new_last_col_id < meta.last_column_id) {
+                vlog(
+                  log.error,
+                  "Expected a highest last column id, {} < {}",
```

Comment:
This error message is a little unclear. "Expected new last column id to be greater than previous last column id: {} < {}" ?

The error message below for sequence numbers is much clearer: "New sequence number must be higher than last: {} <= {}"

Reply:
Makes sense. Will do in a followup